File: crates/pk-core/src/maze/relay_chambers.rs
//! Relay Chambers — two-focus ellipse bank arcs connecting room doorway mouths.
//!
//! PORTS: `maze/relay-chambers.ts`

use std::collections::HashMap;
use std::f64::consts::PI;

use crate::flow_field::bfs_distances;
use crate::grid::{ensure_arcs, idx, is_walkable, Grid};
use crate::maze::arc_contract::junction_clear;
use crate::maze::conic_fit::{arc_chain_from_samples, ellipse_from_foci, ellipse_samples, Pt};
use crate::maze::doorway_funnels::{
    commit_jaw, lane_toward_mouth, plan_chain, revert_jaw, JawPlan,
};
use crate::maze::doorways::Doorway;
use crate::maze::track_launch::TilePos;
use crate::tile_shape::ArcFeature;

pub const RELAY_STANDOFF: f64 = 1.6;
pub const RELAY_MIN_SPAN: f64 = 5.0;
pub const RELAY_MAX_SPAN: f64 = 26.0;
pub const RELAY_SEGMENTS: usize = 5;
pub const RELAY_MAX_PER_FLOOR: usize = 3;

#[derive(Debug, Clone, Default)]
pub struct RelayTune {
    pub standoff: Option<f64>,
    pub segments: Option<usize>,
    pub max_per_floor: Option<usize>,
}

#[derive(Debug, Clone, Default)]
pub struct RelayReport {
    pub chambers: usize,
    pub features: usize,
    pub carved: usize,
    pub filled: usize,
    pub reverted: usize,
    pub rejects: HashMap<String, usize>,
}

fn mouth(d: &Doorway) -> Pt {
    Pt {
        x: d.site.i as f64 + 0.5,
        z: d.site.j as f64 + 0.5,
    }
}

pub fn author_relay_chambers<F: Fn(i32, i32) -> bool>(
    g: &mut Grid,
    doorways: &[Doorway],
    start: TilePos,
    occupied: F,
    tune: &RelayTune,
) -> RelayReport {
    let mut report = RelayReport::default();
    if doorways.len() < 2 {
        return report;
    }
    ensure_arcs(g);

    let standoff = tune.standoff.unwrap_or(RELAY_STANDOFF);
    let segments = tune.segments.unwrap_or(RELAY_SEGMENTS);
    let cap = tune.max_per_floor.unwrap_or(RELAY_MAX_PER_FLOOR);

    struct Pair<'a> {
        a: &'a Doorway,
        b: &'a Doorway,
        span: f64,
    }

    let mut pairs = Vec::new();
    for i in 0..doorways.len() {
        for j in (i + 1)..doorways.len() {
            let a = &doorways[i];
            let b = &doorways[j];
            let shares = a.site.a == b.site.a
                || a.site.a == b.site.b
                || a.site.b == b.site.a
                || a.site.b == b.site.b;
            if !shares {
                continue;
            }
            let span = ((a.site.i - b.site.i) as f64).hypot((a.site.j - b.site.j) as f64);
            if span < RELAY_MIN_SPAN || span > RELAY_MAX_SPAN {
                continue;
            }
            pairs.push(Pair { a, b, span });
        }
    }

    pairs.sort_by(|p, q| {
        q.span
            .partial_cmp(&p.span)
            .unwrap_or(std::cmp::Ordering::Equal)
            .then_with(|| p.a.site.i.cmp(&q.a.site.i))
            .then_with(|| p.a.site.j.cmp(&q.a.site.j))
    });

    let mut committed: Vec<JawPlan> = Vec::new();
    let mut built = 0;

    for pair in pairs {
        if report.chambers >= cap {
            break;
        }
        let f1 = mouth(pair.a);
        let f2 = mouth(pair.b);
        let c = (f2.x - f1.x).hypot(f2.z - f1.z) / 2.0;
        let e = ellipse_from_foci(f1, f2, c + standoff);
        let Some(e) = e else {
            continue;
        };

        let mut halves = 0;
        for (t0, t1) in [(0.12, PI - 0.12), (PI + 0.12, 2.0 * PI - 0.12)] {
            let chain = arc_chain_from_samples(&ellipse_samples(&e, t0, t1, segments), true, Some("funnel"));
            if chain.is_empty() {
                continue;
            }
            let plan = plan_chain(g, &chain, &occupied, &|_, _| false);
            let Ok(mut plan) = plan else {
                if let Err(rej) = plan {
                    *report.rejects.entry(format!("{rej:?}")).or_insert(0) += 1;
                }
                continue;
            };

            let all_clear = plan.features.iter().enumerate().all(|(k, f)| {
                let coords: Vec<(i32, i32)> = plan.arc_tiles[k].iter().map(|t| (t.i, t.j)).collect();
                junction_clear(g, &coords, f)
            });
            if !all_clear {
                *report.rejects.entry("junction".to_string()).or_insert(0) += 1;
                continue;
            }

            plan.mouth = None;
            for f in &mut plan.features {
                let mid = f.a0 + f.span / 2.0;
                let px = f.cx + mid.cos() * f.r;
                let pz = f.cz + mid.sin() * f.r;
                let near = if (px - f1.x).hypot(pz - f1.z) <= (px - f2.x).hypot(pz - f2.z) {
                    f2
                } else {
                    f1
                };
                f.lanes = vec![lane_toward_mouth(f, near)];
            }
            built += commit_jaw(g, &mut plan);
            report.carved += plan.carve_tiles.len();
            report.filled += plan.fill_tiles.len();
            committed.push(plan);
            halves += 1;
        }
        if halves > 0 {
            report.chambers += 1;
        }
    }

    let stranded = |grid: &Grid| -> bool {
        let d = bfs_distances(grid, start.i, start.j);
        for j in 0..grid.h {
            for i in 0..grid.w {
                if is_walkable(grid, i, j) && d[idx(grid, i, j)] < 0 {
                    return true;
                }
            }
        }
        false
    };

    while !committed.is_empty() && stranded(g) {
        let mut p = committed.pop().unwrap();
        revert_jaw(g, &mut p);
        built -= p.features.len();
        report.reverted += p.features.len();
        report.carved -= p.carve_tiles.len();
        report.filled -= p.fill_tiles.len();
    }
    report.features = built;
    report
}

pub fn relay_ellipse(a: &Doorway, b: &Doorway, standoff: f64) -> Vec<ArcFeature> {
    let f1 = mouth(a);
    let f2 = mouth(b);
    let c = (f2.x - f1.x).hypot(f2.z - f1.z) / 2.0;
    let e = ellipse_from_foci(f1, f2, c + standoff);
    let Some(e) = e else {
        return Vec::new();
    };
    arc_chain_from_samples(&ellipse_samples(&e, 0.12, PI - 0.12, RELAY_SEGMENTS), true, Some("funnel"))
}

#[derive(Debug, Clone, PartialEq)]
pub struct RelayEllipse {
    pub f1: (f64, f64),
    pub f2: (f64, f64),
    pub center: (f64, f64),
    pub a: f64,
    pub b: f64,
    pub c: f64,
    pub angle: f64,
}

impl RelayEllipse {
    pub fn point_at(&self, theta: f64) -> (f64, f64) {
        let x_local = self.a * theta.cos();
        let y_local = self.b * theta.sin();

        let cos_a = self.angle.cos();
        let sin_a = self.angle.sin();

        let wx = self.center.0 + x_local * cos_a - y_local * sin_a;
        let wz = self.center.1 + x_local * sin_a + y_local * cos_a;

        (wx, wz)
    }

    pub fn normal_at(&self, theta: f64) -> (f64, f64) {
        let nx_local = theta.cos() / self.a;
        let ny_local = theta.sin() / self.b;
        let len = (nx_local * nx_local + ny_local * ny_local).sqrt();
        let (nx_unit, ny_unit) = (nx_local / len, ny_local / len);

        let cos_a = self.angle.cos();
        let sin_a = self.angle.sin();

        let wx = nx_unit * cos_a - ny_unit * sin_a;
        let wz = nx_unit * sin_a + ny_unit * cos_a;

        (wx, wz)
    }
}

pub fn compute_relay_ellipse(
    d1: (f64, f64),
    d2: (f64, f64),
    standoff: f64,
) -> Option<RelayEllipse> {
    let dx = d2.0 - d1.0;
    let dz = d2.1 - d1.1;
    let focal_dist = dx.hypot(dz);
    let c = focal_dist * 0.5;

    if standoff <= 0.0 {
        return None;
    }

    let a = c + standoff;
    let b = (a * a - c * c).sqrt();
    let center = ((d1.0 + d2.0) * 0.5, (d1.1 + d2.1) * 0.5);
    let angle = dz.atan2(dx);

    Some(RelayEllipse {
        f1: d1,
        f2: d2,
        center,
        a,
        b,
        c,
        angle,
    })
}

pub fn sample_relay_arc(
    ellipse: &RelayEllipse,
    theta_start: f64,
    theta_end: f64,
    num_samples: usize,
) -> Vec<(f64, f64)> {
    if num_samples < 2 {
        return Vec::new();
    }

    let mut samples = Vec::with_capacity(num_samples);
    let step = (theta_end - theta_start) / (num_samples - 1) as f64;

    for i in 0..num_samples {
        let theta = theta_start + step * i as f64;
        samples.push(ellipse.point_at(theta));
    }

    samples
}

File: ThreeJS/src/game/pinball-knight/maze/relay-chambers.ts
/**
 * RELAY CHAMBERS — the two-focus ellipse, and the strongest form of the ask.
 *
 * ── What an ellipse buys that a parabola cannot ──────────────────────────
 *
 * A parabola gathers rays that arrive PARALLEL to its axis. That is the right
 * tool for a corridor, and it is what `doorway-funnels.ts` uses. But it says
 * nothing about a ball crossing an open section on some other line.
 *
 * An ellipse's property is stronger and stranger: every ray through focus F₁
 * reflects to focus F₂, whatever direction it left F₁ in. Put the two foci on
 * TWO DOORWAY MOUTHS of the same section and the precondition stops being an
 * assumption — a ball that entered through door A has, by construction, just
 * passed through F₁. So the far wall relays it to door B in a single bank, from
 * any angle it happened to come in at.
 *
 * That is "it banks into the next room no matter how you go in", holding
 * exactly rather than approximately, and it is the reason to want ellipses
 * rather than just curves.
 *
 * ── Why it is authored as a WALL and not a room ──────────────────────────
 *
 * The obvious build is a chamber shaped like an ellipse. There is nowhere to
 * put one: `LONG_BANKS_PLAN.md` censused 22,713 open tiles and found 81.8% with
 * an open radius of ZERO. Floors are corridors.
 *
 * So this authors only the ARC — the stretch of the section's existing boundary
 * that happens to lie near the ellipse — and lets the rest of the section stay
 * whatever shape it is. Cost is perimeter, not area, which is the same reframe
 * that made artery banks affordable. A partial relay wall is still a relay
 * wall: the balls that meet it get sent to the far door, and the ones that miss
 * it are no worse off than before.
 *
 * ── Siting ───────────────────────────────────────────────────────────────
 *
 * `Doorway` already names the two SECTIONS it joins, so a pair of doorways that
 * share a section is a relay candidate with no geometry needed to find it —
 * censused at 10.0 pairs per floor, of which 5.4 have the standoff to fit an
 * ellipse. The semi-major axis is `c + RELAY_STANDOFF`, which is the one free
 * parameter: it decides how far off the line between the mouths the wall sits.
 *
 * Reuses `doorway-funnels.ts` wholesale for the parts that are not about
 * ellipses — tile classification, the fill/carve rules, the snapshot revert and
 * the collective strand guard — because those were expensive to get right and a
 * second copy would drift. DOM- and three-free. Pure.
 */
import { type Grid, type TilePos, idx, isWalkable, ensureArcs } from "./generator";
import type { ArcFeature } from "../engine/tile-shape";
import { junctionClear } from "./arc-contract";
import { bfsDistancesOwned } from "../engine/flow-field";
import { ellipseFromFoci, ellipseSamples, arcChainFromSamples, type Pt } from "./conic-fit";
import {
  planChain,
  commitJaw,
  revertJaw,
  laneTowardMouth,
  type JawPlan,
  type Occupied,
} from "./doorway-funnels";
import type { Doorway } from "./doorways";

/**
 * How far outside the straight line between the two mouths the relay wall sits,
 * in tiles — the ellipse's semi-major axis is `c + this`.
 *
 * Small values make a long thin ellipse hugging the line between the doors,
 * which is where a corridor's walls actually are. Large values need a chamber,
 * and there are none.
 */
export const RELAY_STANDOFF = 1.6;

/** Mouth separations worth relaying, in tiles. */
export const RELAY_MIN_SPAN = 5;
export const RELAY_MAX_SPAN = 26;

/** Arc links per relay wall. */
export const RELAY_SEGMENTS = 5;

/** Relay walls per floor. */
export const RELAY_MAX_PER_FLOOR = 3;

export interface RelayReport {
  chambers: number;
  features: number;
  carved: number;
  filled: number;
  reverted: number;
  rejects: Record<string, number>;
}

const mouth = (d: Doorway): Pt => ({ x: d.i + 0.5, z: d.j + 0.5 });

/**
 * Author elliptical relay walls between doorway pairs that share a section.
 *
 * Both halves of each ellipse are tried — the wall may exist on either side of
 * the line between the mouths, and on a corridor it is usually one but not the
 * other. Each is planned and gated independently, so half a relay is a normal
 * outcome here rather than the failure it is for a funnel jaw. The difference
 * is that a funnel's two arms face EACH OTHER across a throat, so a lone arm
 * deflects into the space its partner should have filled; two halves of an
 * ellipse face the same two foci and neither needs the other to aim correctly.
 */
export function authorRelayChambers(
  g: Grid,
  doorways: readonly Doorway[],
  start: TilePos,
  occupied: Occupied = () => false,
  tune: { standoff?: number; segments?: number; maxPerFloor?: number } = {},
): RelayReport {
  const report: RelayReport = { chambers: 0, features: 0, carved: 0, filled: 0, reverted: 0, rejects: {} };
  if (doorways.length < 2) return report;
  ensureArcs(g);

  const standoff = tune.standoff ?? RELAY_STANDOFF;
  const segments = tune.segments ?? RELAY_SEGMENTS;
  const cap = tune.maxPerFloor ?? RELAY_MAX_PER_FLOOR;

  // Pairs that share a section, in a fixed order — this runs inside
  // `buildTrackFloor`'s single seeded stream and must not draw from it.
  const pairs: Array<{ a: Doorway; b: Doorway; span: number }> = [];
  for (let i = 0; i < doorways.length; i++) {
    for (let j = i + 1; j < doorways.length; j++) {
      const a = doorways[i];
      const b = doorways[j];
      const shares = a.a === b.a || a.a === b.b || a.b === b.a || a.b === b.b;
      if (!shares) continue;
      const span = Math.hypot(a.i - b.i, a.j - b.j);
      if (span < RELAY_MIN_SPAN || span > RELAY_MAX_SPAN) continue;
      pairs.push({ a, b, span });
    }
  }
  // Widest first: a long relay is a longer ride and a more legible object.
  pairs.sort((p, q) => q.span - p.span || p.a.i - q.a.i || p.a.j - q.a.j);

  const committed: JawPlan[] = [];
  let built = 0;

  for (const { a, b } of pairs) {
    if (report.chambers >= cap) break;
    const f1 = mouth(a);
    const f2 = mouth(b);
    const c = Math.hypot(f2.x - f1.x, f2.z - f1.z) / 2;
    const e = ellipseFromFoci(f1, f2, c + standoff);
    if (!e) continue;

    let halves = 0;
    // The two halves of the ellipse, either side of the major axis.
    for (const [t0, t1] of [
      [0.12, Math.PI - 0.12],
      [Math.PI + 0.12, 2 * Math.PI - 0.12],
    ] as const) {
      const chain = arcChainFromSamples(ellipseSamples(e, t0, t1, segments), true, "funnel");
      if (chain.length === 0) continue;
      // Nothing is sealed here: a relay wall never sits in a doorway's own
      // threshold, because it is authored along the boundary BETWEEN the two
      // mouths rather than across either of them.
      const plan = planChain(g, chain, occupied, () => false);
      if (typeof plan === "string") {
        report.rejects[plan] = (report.rejects[plan] ?? 0) + 1;
        continue;
      }
      if (!plan.features.every((f, k) => junctionClear(g, plan.arcTiles[k], f))) {
        report.rejects.junction = (report.rejects.junction ?? 0) + 1;
        continue;
      }
      // A relay wall is ridden toward whichever mouth is nearer the stretch
      // that survived, so the lane is aimed per-feature rather than per-wall.
      plan.mouth = undefined;
      for (const f of plan.features) {
        const mid = f.a0 + f.span / 2;
        const px = f.cx + Math.cos(mid) * f.r;
        const pz = f.cz + Math.sin(mid) * f.r;
        const near = Math.hypot(px - f1.x, pz - f1.z) <= Math.hypot(px - f2.x, pz - f2.z) ? f2 : f1;
        // Carry toward the FAR mouth: the near one is where the ball came from.
        f.lanes = [laneTowardMouth(f, near)];
      }
      built += commitJaw(g, plan);
      committed.push(plan);
      report.carved += plan.carveTiles.length;
      report.filled += plan.fillTiles.length;
      halves++;
    }
    if (halves > 0) report.chambers++;
  }

  // Same collective guard as the funnels, unwound one wall at a time — see
  // `doorway-funnels.ts` for why it is unconditional and why it is not
  // all-or-nothing.
  const stranded = (): boolean => {
    const d = bfsDistancesOwned(g, start.i, start.j);
    for (let j = 0; j < g.h; j++) {
      for (let i = 0; i < g.w; i++) {
        if (isWalkable(g, i, j) && d[idx(g, i, j)] < 0) return true;
      }
    }
    return false;
  };
  while (committed.length > 0 && stranded()) {
    const p = committed.pop()!;
    revertJaw(g, p);
    built -= p.features.length;
    report.reverted += p.features.length;
    report.carved -= p.carveTiles.length;
    report.filled -= p.fillTiles.length;
  }
  report.features = built;
  return report;
}

/** The ellipse a pair of mouths would relay across — exposed for tests. */
export function relayEllipse(a: Doorway, b: Doorway, standoff = RELAY_STANDOFF): ArcFeature[] {
  const f1 = mouth(a);
  const f2 = mouth(b);
  const c = Math.hypot(f2.x - f1.x, f2.z - f1.z) / 2;
  const e = ellipseFromFoci(f1, f2, c + standoff);
  if (!e) return [];
  return arcChainFromSamples(ellipseSamples(e, 0.12, Math.PI - 0.12, RELAY_SEGMENTS), true, "funnel");
}
